presto-main/src/test/java/com/facebook/presto/operator/scalar/TestHyperLogLogFunctions.java

```diff
@@ -0,0 +1,147 @@
+    }
+
+    @Test
+    public void testCardinalitySingleColumn()
+    {
+        long blockSize = 1;
+        long uniqueElts = 10000 * blockSize;
+
+        String projection = getCardinalityProjection(getUniqueElements((int) blockSize, uniqueElts));
+
+        double error = uniqueElts * 0.05;
+        functionAssertions.assertFunctionWithError(projection, BIGINT, uniqueElts, error);
+    }
+
+    @Test
+    public void testCardinalityTwoColumns()
+    {
+        long blockSize = 2;
+        long uniqueElts = 10000 * blockSize;
+
+        String projection = getCardinalityProjection(getUniqueElements((int) blockSize, uniqueElts));
+
+        double error = uniqueElts * 0.05;
+        functionAssertions.assertFunctionWithError(projection, BIGINT, uniqueElts, error);
+    }
+
+    @Test
+    public void testCardinalityThreeColumns()
+    {
+        long blockSize = 3;
+        long uniqueElts = 10000 * blockSize;
+
+        String projection = getCardinalityProjection(getUniqueElements((int) blockSize, uniqueElts));
+
+        double error = uniqueElts * 0.05;
+        functionAssertions.assertFunctionWithError(projection, BIGINT, uniqueElts, error);
+    }
+
+    @Test
+    public void testCardinalityManyColumns()
+    {
+        // max number of columns to merge is 254
+        long blockSize = 254;
+        long uniqueElts = 1000 * blockSize;
+        String projection = getCardinalityProjection(getUniqueElements((int) blockSize, uniqueElts));
+
+        double error = uniqueElts * 0.05;
+        functionAssertions.assertFunctionWithError(projection, BIGINT, uniqueElts, error);
+    }
+
+    private HyperLogLog[] getUniqueElements(int blockSize, long uniqueElements)
+    {
+        HyperLogLog[] array = new HyperLogLog[blockSize];
+
+        for (int j = 0; j < blockSize; j++) {
+            // create a single HyperLogLog column
+            HyperLogLog firstHll = HyperLogLog.newInstance(NUMBER_OF_BUCKETS);
+
+            // populate column with even partitions of the unique elements
+            for (long i = j * uniqueElements / blockSize; i < j * uniqueElements / blockSize +
+                    uniqueElements / blockSize; i++) {
+                firstHll.add(i);
+            }
+
+            array[j] = firstHll;
+        }
+
+        return array;
+    }
+
+    private String getCardinalityProjection(HyperLogLog... array)
+    {
+        String projection = "cardinality(merge_hll(ARRAY[";
+
+        for (int j = 0; j < array.length; j++) {
+            // serialize the HyperLogLog to a Slice
+            Slice firstSerial = array[j].serialize();
+
+            // get bytes of Slice
+            byte[] firstBytes = firstSerial.getBytes();
+
+            // convert bytes into String
+            String firstEncode = BaseEncoding.base16().lowerCase().encode(firstBytes);
+
+            // write our projection statement
+            projection += "CAST(X'" + firstEncode + "' AS HyperLogLog)";
+            if (j < array.length - 1) {
```

Comment:
Might be cleaner to construct the list of CASTs and then use Joiner to join.

https://google.github.io/guava/releases/27.1-jre/api/docs/com/google/common/base/Joiner.html

`Joiner.on(", ").join(myListOfCastStatementsAsString);`